### PR TITLE
Set device before calling cudaStreamWaitEvent() in CUDA IPC channel.

### PR DESCRIPTION
**Pull request by beauby**
Summary: A call to cudaStreamWaitEvent() should happen on the device the stream is defined on, according to CUDA docs.

Differential Revision: D24859542

### CODE REVIEW

File: tensorpipe/test/test_environment.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#if TENSORPIPE_SUPPORTS_CUDA
+#include <tensorpipe/common/cuda.h>
```
Comment (lw) [inline tensorpipe/test/test_environment.h:12]: I don't think this header is needed? Maybe move it to the .cc file?

File: tensorpipe/test/test_environment.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <tensorpipe/common/cuda.h>
+#include <tensorpipe/common/defs.h>
+#include <unistd.h>
+#endif // TENSORPIPE_SUPPORTS_CUDA
+
+int TestEnvironment::numCudaDevices() {
+  static int count = -1;
+  if (count == -1) {
```
Comment (lw) [inline tensorpipe/test/test_environment.cc:24]: Do we need to worry about race conditions when multiple threads call this function at the same time? (The C++ runtime guarantees atomicity on the assignment of a local static variable but not on subsequent modifications).

File: tensorpipe/test/test_environment.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+#endif // TENSORPIPE_SUPPORTS_CUDA
+
+class TestEnvironment {
+ public:
+  static int numCudaDevices();
```
Comment (lw) [inline tensorpipe/test/test_environment.h:17]: I'm wondering if we should rather make this a member function (a method) rather than a static one. For now a static function makes sense, so that we don't have to pass an instance of the environment around everywhere, but perhaps one day once we add more fields to this we'll want to run a test multiple time with different environments. (In general I guess this relates to my tendency to avoid global state as much as possible).

TL;DR: this is perfectly fine for now, but there's food for thought.

File: tensorpipe/test/channel/channel_test_cuda.cc [hunk 4]
```diff
@@ -120,3 +121,118 @@ class ReceiverWaitsForStartEventTest
...
 CHANNEL_TEST(CudaChannelTestSuite, ReceiverWaitsForStartEvent);
+
+class SendAcrossDevicesTest : public ClientServerChannelTestCase<CudaBuffer> {
+  static constexpr size_t kSize = 1024;
+
+ public:
+  void run(ChannelTestHelper<CudaBuffer>* helper) override {
+    if (TestEnvironment::numCudaDevices() < 2) {
+      GTEST_SKIP() << "Skipping test requiring >=2 CUDA devices.";
+      return;
```
Comment (lw) [inline tensorpipe/test/channel/channel_test_cuda.cc:132]: Do we need to return? One of the gtest examples seems to hint GTEST_SKIP will somehow skips what comes after (I guess it raises an exception). https://github.com/google/googletest/blob/master/googletest/test/gtest_skip_test.cc#L37-L40

**Review by lw — APPROVED** [review]
Thanks for adding a test (and the infrastructure needed to get it to work) and to make it harder to forget this in the future!

CircleCI "correctly" skips the test, but I'm assuming you ran it for both XTH and IPC manually to ensure it passes?

File: tensorpipe/test/test_environment.cc (hunk 2 above)
Comment (beauby) [reply tensorpipe/test/test_environment.cc:24]: The race condition here is pretty harmless, as `cudaGetDeviceCount()` will always return the same value, so worst case it will be called multiple times.

File: tensorpipe/test/test_environment.h (hunk 1 above)
Comment (beauby) [reply tensorpipe/test/test_environment.h:12]: Good catch, removed.

File: tensorpipe/test/channel/channel_test_cuda.cc (hunk 4 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test_cuda.cc:132]: Updated.
